subprojects/snapshots/src/test/groovy/org/gradle/internal/snapshot/AbstractSnapshotWithChildrenTest.groovy

```diff
@@ -0,0 +1,208 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot
+
+import spock.lang.Specification
+
+import javax.annotation.Nullable
+import java.util.stream.Collectors
+
+import static org.gradle.internal.snapshot.CaseSensitivity.CASE_SENSITIVE
+
+abstract class AbstractSnapshotWithChildrenTest<T extends FileSystemNode, CHILD extends FileSystemNode> extends Specification {
+    T initialRoot
+    List<CHILD> children
+    String absolutePath
+    int offset
+    FileSystemNode selectedChild
+
+    abstract protected T createInitialRootNode(String pathToParent, List<CHILD> children);
+
+    abstract protected CHILD mockChild(String pathToParent)
+
+    void setupTest(VirtualFileSystemTestSpec spec) {
+        this.children = createChildren(spec.childPaths)
+        this.initialRoot = createInitialRootNode("path/to/parent", children)
+        this.absolutePath = spec.absolutePath
+        this.offset = spec.offset
+        this.selectedChild = children.find { it.pathToParent == spec.selectedChildPath }
+    }
+
+    List<CHILD> createChildren(List<String> pathsToParent) {
+        return pathsToParent.stream()
+            .sorted(PathUtil.getPathComparator(CASE_SENSITIVE))
+            .map { childPath -> mockChild(childPath) }
+            .collect(Collectors.toList())
+    }
+
+    List<FileSystemNode> childrenWithSelectedChildReplacedBy(FileSystemNode replacement) {
+        children.collect { it == selectedChild ? replacement : it }
+    }
+
+    List<FileSystemNode> childrenWithAdditionalChild(FileSystemNode newChild) {
+        List<FileSystemNode> newChildren = new ArrayList<>(children)
+        newChildren.add(insertionPoint, newChild)
+        return newChildren
+    }
+
+    List<CHILD> childrenWithSelectedChildRemoved() {
+        children.findAll { it != selectedChild }
+    }
+
+    CHILD getNodeWithIndexOfSelectedChild(List<CHILD> newChildren) {
+        int index = children.indexOf(selectedChild)
+        return newChildren[index]
+    }
+
+    private int getInsertionPoint() {
+        int childIndex = SearchUtil.binarySearch(children) {
+            candidate -> PathUtil.compareFirstSegment(absolutePath, offset, candidate.pathToParent, CASE_SENSITIVE)
+        }
+        assert childIndex < 0
+        return -childIndex - 1
+    }
+
+    String getCommonPrefix() {
+        return selectedChild.pathToParent.substring(0, PathUtil.lengthOfCommonPrefix(selectedChild.pathToParent, absolutePath, offset, CASE_SENSITIVE))
+    }
+
+    String getPathFromCommonPrefix() {
+        return absolutePath.substring(offset + commonPrefix.length() + 1)
+    }
+
+    String getSelectedChildPathFromCommonPrefix() {
+        return selectedChild.pathToParent.substring(commonPrefix.length() + 1)
+    }
```

Comment:
This string index magic is still confusing for me. Let's merge it if it works, and maybe we can do another round of review for https://github.com/gradle/gradle/pull/11378.

Reply:
👍